File: upgrade.py
import json
import os
import platform
import stat
import sys
import tempfile
from urllib.error import HTTPError
from urllib.request import Request, urlopen

GITHUB_REPO = "omertahaoztop/vikunja-tui"
ASSET_NAME = "vikunja-tui-linux-amd64"


def _github_get(path: str) -> dict:
    url = f"https://api.github.com{path}"
    req = Request(url, headers={"Accept": "application/vnd.github.v3+json"})
    with urlopen(req) as r:
        result = json.loads(r.read())
    if not isinstance(result, dict):
        raise RuntimeError(f"Unexpected response from {path}")
    return result


def _get_current_binary() -> str | None:
    if getattr(sys, "frozen", False):
        return os.path.realpath(sys.executable)
    return None


def self_upgrade(current_version: str) -> None:
    binary_path = _get_current_binary()
    if not binary_path:
        print("Not running as a compiled binary. Use git pull to update from source.")
        sys.exit(1)

    if platform.system() != "Linux" or platform.machine() not in ("x86_64", "AMD64"):
        print("Self-upgrade is only supported on Linux amd64.")
        sys.exit(1)

    print("Checking for updates...")

    try:
        release = _github_get(f"/repos/{GITHUB_REPO}/releases/latest")
    except HTTPError as e:
        print(f"Failed to check for updates: HTTP {e.code}")
        sys.exit(1)
    except Exception as e:
        print(f"Failed to check for updates: {e}")
        sys.exit(1)

    tag = release["tag_name"]

    if current_version != "dev" and current_version == tag:
        print(f"Already up to date ({tag}).")
        sys.exit(0)

    asset_url = None
    for asset in release.get("assets", []):
        if asset["name"] == ASSET_NAME:
            asset_url = asset["browser_download_url"]
            break

    if not asset_url:
        print(f"No compatible binary found in release {tag}.")
        sys.exit(1)

    if current_version == "dev":
        print(f"Downloading latest release ({tag})...")
    else:
        print(f"Updating {current_version} -> {tag}...")

    try:
        req = Request(asset_url)
        with urlopen(req) as r:
            data = r.read()
    except Exception as e:
        print(f"Download failed: {e}")
        sys.exit(1)

    binary_dir = os.path.dirname(binary_path)
    tmp_path: str | None = None
    try:
        fd, tmp_path = tempfile.mkstemp(dir=binary_dir, prefix=".vikunja-tui-upgrade-")
        os.write(fd, data)
        os.close(fd)
        original_mode = os.stat(binary_path).st_mode
        os.chmod(tmp_path, original_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
        os.replace(tmp_path, binary_path)
    except PermissionError:
        if tmp_path:
            try:
                os.unlink(tmp_path)
            except OSError:
                pass
        print(f"Permission denied. Try:\n  sudo {binary_path} --upgrade")
        sys.exit(1)
    except Exception:
        if tmp_path:
            try:
                os.unlink(tmp_path)
            except OSError:
                pass
        raise

    print(f"Updated to {tag}.")

File: internal/upgrade/upgrade.go
package upgrade

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"time"
)

const (
	githubRepo = "omertahaoztop/vikunja-tui"
	assetName  = "vikunja-tui-linux-amd64"
)

type release struct {
	TagName string `json:"tag_name"`
	Assets  []struct {
		Name string `json:"name"`
		URL  string `json:"browser_download_url"`
	} `json:"assets"`
}

func githubGet(path string, out any) error {
	req, err := http.NewRequest(http.MethodGet, "https://api.github.com"+path, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	c := &http.Client{Timeout: 30 * time.Second}
	resp, err := c.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	return json.Unmarshal(raw, out)
}

func download(url string) ([]byte, error) {
	c := &http.Client{Timeout: 5 * time.Minute}
	resp, err := c.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("download HTTP %d", resp.StatusCode)
	}
	return io.ReadAll(resp.Body)
}

func SelfUpgrade(currentVersion string) error {
	if runtime.GOOS != "linux" || runtime.GOARCH != "amd64" {
		return fmt.Errorf("self-upgrade is only supported on linux/amd64")
	}

	binaryPath, err := os.Executable()
	if err != nil {
		return fmt.Errorf("cannot locate binary: %w", err)
	}
	binaryPath, _ = filepath.EvalSymlinks(binaryPath)

	fmt.Println("Checking for updates...")
	var rel release
	if err := githubGet("/repos/"+githubRepo+"/releases/latest", &rel); err != nil {
		return fmt.Errorf("failed to check for updates: %w", err)
	}

	if currentVersion != "dev" && currentVersion == rel.TagName {
		fmt.Printf("Already up to date (%s).\n", rel.TagName)
		return nil
	}

	var assetURL string
	for _, a := range rel.Assets {
		if a.Name == assetName {
			assetURL = a.URL
			break
		}
	}
	if assetURL == "" {
		return fmt.Errorf("no compatible binary found in release %s", rel.TagName)
	}

	if currentVersion == "dev" {
		fmt.Printf("Downloading latest release (%s)...\n", rel.TagName)
	} else {
		fmt.Printf("Updating %s -> %s...\n", currentVersion, rel.TagName)
	}

	data, err := download(assetURL)
	if err != nil {
		return fmt.Errorf("download failed: %w", err)
	}

	dir := filepath.Dir(binaryPath)
	tmp, err := os.CreateTemp(dir, ".vikunja-tui-upgrade-")
	if err != nil {
		return fmt.Errorf("permission denied. Try:\n  sudo %s --upgrade", binaryPath)
	}
	tmpPath := tmp.Name()

	cleanup := func() { _ = os.Remove(tmpPath) }
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		cleanup()
		return err
	}
	tmp.Close()

	info, err := os.Stat(binaryPath)
	if err == nil {
		_ = os.Chmod(tmpPath, info.Mode()|0o111)
	} else {
		_ = os.Chmod(tmpPath, 0o755)
	}

	if err := os.Rename(tmpPath, binaryPath); err != nil {
		cleanup()
		return fmt.Errorf("permission denied. Try:\n  sudo %s --upgrade", binaryPath)
	}

	fmt.Printf("Updated to %s.\n", rel.TagName)
	return nil
}
